File: Samples/C#/GeometryGridTypes/GeometryGridTypesPlugin.cs
/*  Sample: Geometry Grid Types
 *
 *  This sample illustrates how to work with export object geometry grid types via the Renga API.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

using System;
using System.Text;

namespace GeometryGridTypes
{
  public class GeometryGridTypesPlugin : Renga.IPlugin
  {
    private Renga.Application m_app;
    private Renga.ActionEventSource m_events;

    public bool Initialize(string pluginFolder)
    {
      m_app = new Renga.Application();
      var ui = m_app.UI;

      var action = ui.CreateAction();
      action.ToolTip = "Export geometry and discover mesh and grid types";

      m_events = new Renga.ActionEventSource(action);
      m_events.Triggered += (s, e) => { ReportSceneContents(); };

      var panelExtension = ui.CreateUIPanelExtension();
      panelExtension.AddToolButton(action);

      ui.AddExtensionToPrimaryPanel(panelExtension);
      return true;
    }

    public void Stop()
    {
      m_events.Dispose();
    }


    private void ReportSceneContents()
    {
      var project = m_app.Project;
      var dataExporter = project.DataExporter;

      var stringBuilder = new StringBuilder();
      gatherExportData(stringBuilder, dataExporter.GetObjects3D());

      m_app.UI.ShowMessageBox(Renga.MessageIcon.MessageIcon_Info,
                                   "Result of exporting the 3D objects",
                                   stringBuilder.ToString());
    }

    private void gatherExportData(StringBuilder builder, Renga.IExportedObject3DCollection object3DCollection)
    {
      builder.AppendLine("Exported objects:");
      builder.AppendLine();

      int size = object3DCollection.Count;
      for (int i = 0; i < size; ++i)
        gatherExportData(builder, object3DCollection.Get(i));
    }

    private void gatherExportData(StringBuilder builder, Renga.IExportedObject3D object3D)
    {
      var id = object3D.ModelObjectId;
      var modelObject = m_app.Project.Model.GetObjects().GetById(id);

      builder.Append("Object: ");
      builder.AppendLine(modelObject?.Name ?? "undefined model object");

      var meshCount = object3D.MeshCount;
      for (int i = 0; i < meshCount; ++i)
        gatherExportData(builder, object3D.GetMesh(i), object3D.ModelObjectType);
    }

    private void gatherExportData(StringBuilder builder, Renga.IMesh mesh, Guid objectTypeGuid)
    {
      builder.Append("Mesh: type - ");

      var meshType = mesh.MeshType;
      if (meshType == Renga.MeshTypes.Undefined)
        builder.AppendLine("undefined");
      else if (meshType == Renga.MeshTypes.DoorReveal)
        builder.AppendLine("door reveal");
      else if (meshType == Renga.MeshTypes.DoorPanel)
        builder.AppendLine("door panel");
      else if (meshType == Renga.MeshTypes.DoorTransom)
        builder.AppendLine("door transom");
      else if (meshType == Renga.MeshTypes.DoorLining)
		builder.AppendLine("door lining");
      else if (meshType == Renga.MeshTypes.DoorThreshold)
		builder.AppendLine("door threshold");
      else if (meshType == Renga.MeshTypes.DoorPlatband)
	    builder.AppendLine("door platband");
      else if (meshType == Renga.MeshTypes.WindowReveal)
        builder.AppendLine("window reveal");
      else if (meshType == Renga.MeshTypes.WindowPanel)
        builder.AppendLine("window panel");
      else if (meshType == Renga.MeshTypes.WindowSill)
        builder.AppendLine("window sill");
      else if (meshType == Renga.MeshTypes.WindowOutwardSill)
        builder.AppendLine("window outward sill");

      var gridCount = mesh.GridCount;
      for (int i = 0; i < gridCount; ++i)
        gatherExportData(builder, mesh.GetGrid(i), objectTypeGuid);
    }

    void gatherExportData(StringBuilder builder, Renga.IGrid grid, Guid objectTypeGuid)
    {
      if (objectTypeGuid == Renga.ObjectTypes.Wall)
      {
        builder.AppendLine("  Grid: type - wall ");
        switch ((Renga.GridTypes.Wall)grid.GridType)
        {
        case Renga.GridTypes.Wall.Undefined:
          builder.AppendLine("undefined");
          break;
        case Renga.GridTypes.Wall.FrontSide:
          builder.AppendLine("front side");
          break;
        case Renga.GridTypes.Wall.BackSide:
          builder.AppendLine("back side");
          break;
        case Renga.GridTypes.Wall.Bottom:
          builder.AppendLine("bottom");
          break;
        case Renga.GridTypes.Wall.Top:
          builder.AppendLine("top");
          break;
        case Renga.GridTypes.Wall.FirstButt:
          builder.AppendLine("first butt");
          break;
        case Renga.GridTypes.Wall.SecondButt:
          builder.AppendLine("second butt");
          break;
        case Renga.GridTypes.Wall.Reveal:
          builder.AppendLine("reveal");
          break;
        case Renga.GridTypes.Wall.Cut:
          builder.AppendLine("cut");
          break;
        }
      }
    }
  }
}

File: Samples/C++/GeometryGridTypes/GeometryGridTypesPlugin.cpp
/*  Sample: Geometry Grid Types
 *
 *  This sample illustrates how to work with export object geometry grid types via the Renga API.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */


#include "GeometryGridTypesPlugin.h"

#include <Renga/CreateApplication.hpp>
#include <Renga/MeshTypes.h>
#include <Renga/GridTypes.h>
#include <Renga/ObjectTypes.h>

#include <sstream>


class GridExportHandler : public Renga::ActionEventHandler
{
public:
  GridExportHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
    Renga::ActionEventHandler(pAction),
    m_pApplication(pApplication)
  {
  }

  void OnTriggered() override
  {
    auto pProject = m_pApplication->GetProject();
    auto pDataExporter = pProject->GetDataExporter();
    
    auto command_output = gatherExportData(pDataExporter->GetObjects3D());

    MessageBox(NULL,
      (LPCWSTR)command_output.c_str(),
      (LPCWSTR)L"Export 3D objects result output",
      MB_ICONINFORMATION | MB_OK);
  }

  std::wstring gatherExportData(Renga::IExportedObject3DCollectionPtr pObject3DCollection)
  {
    std::wstring ret;
    ret += L"Objects exported:\n\n";

    auto size = pObject3DCollection->GetCount();
    for (int i = 0; i < size; ++i)
      ret += gatherExportData(pObject3DCollection->Get(i));
    return ret;
  }

  std::wstring gatherExportData(Renga::IExportedObject3DPtr pObject3D)
  {
    std::wstring ret;

    ret += L"Object: ";

    auto id = pObject3D->GetModelObjectId();
    auto pProject = m_pApplication->GetProject();
    auto pModel = pProject->GetModel();
    auto pModelObjectCollection = pModel->GetObjects();
    if (auto pModelObject = pModelObjectCollection->GetById(id))
    {
      ret += pModelObject->GetName();
    }
    else
    {
      ret += L"Undefined as model object";
    }
    ret += L"\n";

    auto meshCount = pObject3D->GetMeshCount();
    for (int i = 0; i < meshCount; ++i)
    {
      ret += gatherExportData(pObject3D->GetMesh(i), pObject3D->GetModelObjectType());
    }

    ret += L"\n";

    return ret;
  }

  std::wstring gatherExportData(Renga::IMeshPtr pMesh, const GUID objectType)
  {
    std::wstring ret;

    ret += L"Mesh: type - ";

    auto meshType = pMesh->GetMeshType();
    if (::IsEqualGUID(meshType, Renga::MeshTypes::Undefined))
      ret += L"undefined";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::DoorReveal))
      ret += L"door reveal";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::DoorPanel))
      ret += L"door panel";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::DoorTransom))
      ret += L"door transom";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::WindowReveal))
      ret += L"window reveal";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::WindowPanel))
      ret += L"window panel";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::WindowSill))
      ret += L"window sill";
    else if (::IsEqualGUID(meshType, Renga::MeshTypes::WindowOutwardSill))
      ret += L"window outward sill";

    ret += L"\n";

    auto gridCount = pMesh->GetGridCount();
    for (int i = 0; i < gridCount; ++i)
      ret += gatherExportData(pMesh->GetGrid(i), objectType);
    return ret;
  }

  std::wstring gatherExportData(Renga::IGridPtr pGrid, const GUID objectType)
  {
    std::wstring ret;
    if (::IsEqualGUID(objectType, Renga::ObjectTypes::Wall))
    {
      ret += L"  Grid: type - wall ";
      switch (pGrid->GetGridType())
      {
      case Renga::GridTypes::Wall::Undefined:
        ret += L"undefined";
        break;
      case Renga::GridTypes::Wall::FrontSide:
        ret += L"front side";
        break;
      case Renga::GridTypes::Wall::BackSide:
        ret += L"back side";
        break;
      case Renga::GridTypes::Wall::Bottom:
        ret += L"bottom";
        break;
      case Renga::GridTypes::Wall::Top:
        ret += L"top";
        break;
      case Renga::GridTypes::Wall::FirstButt:
        ret += L"first butt";
        break;
      case Renga::GridTypes::Wall::SecondButt:
        ret += L"second butt";
        break;
      case Renga::GridTypes::Wall::Reveal:
        ret += L"reveal";
        break;
      case Renga::GridTypes::Wall::Cut:
        ret += L"cut";
        break;
      }
      ret += L"\n";
    }
    return ret;
  }


  void OnToggled(bool checked) override {}

private:
  Renga::IApplicationPtr m_pApplication;
};



GeometryGridTypesPlugin::GeometryGridTypesPlugin()
{
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
}

GeometryGridTypesPlugin::~GeometryGridTypesPlugin()
{
  ::CoUninitialize();
}

bool GeometryGridTypesPlugin::initialize(const wchar_t* pluginDir)
{
  auto pApplication = Renga::CreateApplication();
  if (!pApplication)
    return false;

  if (auto pUI = pApplication->GetUI())
  {
    if (auto pUIPanelExtension = pUI->CreateUIPanelExtension())
    {
      if (auto pAction = pUI->CreateAction())
      {
        pAction->PutToolTip(L"Export geometry and discover mesh and grid types");
        m_handler = std::make_unique<GridExportHandler>(pAction, pApplication);
        pUIPanelExtension->AddToolButton(pAction);
      }
      pUI->AddExtensionToPrimaryPanel(pUIPanelExtension);
    }
  }
  return true;
}

void GeometryGridTypesPlugin::stop()
{
  m_handler.reset();
}
